### PR TITLE
Maintain Referential Integrity while Deleting Resources

### PR DESCRIPTION
**Pull request by alexanderkiel**
Closes: #543

### CODE REVIEW

**Review by alexanderkiel — COMMENTED** [review]
Please also fill out the documentation on the delete transaction command in `docs/implementation/database.md`. Also please write a section on the delete interaction in `docs/api.md` were we say that referential integrity is checked depending on `ENFORCE_REFERENTIAL_INTEGRITY` since version 0.24.0.

File: modules/db/src/blaze/db/impl/index/resource_handle.clj [hunk 1]
```diff
@@ -117,3 +118,6 @@
 
 (defn reference [rh]
   (str (codec/tid->type (tid rh)) "/" (id rh)))
+
+(defn local-ref-tuple [{type :fhir/type id :id}]
```
[the remark this answers is not in the document]
Comment (sonatsuer) [reply modules/db/src/blaze/db/impl/index/resource_handle.clj:122]: Done.

File: modules/db/test/blaze/db/node/tx_indexer/verify_test.clj [hunk 2]
```diff
@@ -29,7 +29,7 @@
    [juxt.iota :refer [given]]
    [taoensso.timbre :as log]))
 
-(st/instrument)
+(prn (st/instrument))
```
[the remark this answers is not in the document]
Comment (sonatsuer) [reply modules/db/test/blaze/db/node/tx_indexer/verify_test.clj]: Done.

File: modules/db/test/blaze/db/node/tx_indexer/verify_test.clj [hunk 3]
```diff
@@ -341,7 +344,49 @@
           [4 1 ss-tu/decode-key] := {:t 2}
           [4 2 ss-tu/decode-val] := {:total 2 :num-changes 2}))))
 
+  (testing "adding an observation referring to an existing patinet"
```
[the remark this answers is not in the document]
Comment (sonatsuer) [reply modules/db/test/blaze/db/node/tx_indexer/verify_test.clj]: Done.

File: modules/db/test/blaze/db/api_test.clj [hunk 4]
```diff
@@ -430,7 +463,45 @@
...
+    (testing "patient with an observation referencing it"
+      (with-system-data [{:blaze.db/keys [node]} config]
+        [[[:create {:fhir/type :fhir/Patient :id "0"}]
+          [:create {:fhir/type :fhir/Observation :id "0"
+                    :subject #fhir/Reference{:reference "Patient/0"}}]]]
+        (testing "deleting only the patient"
+          (given-failed-future
+           (d/transact
+            node
+            [[:delete "Patient" "0"]])
```
[the remark this answers is not in the document]
Comment (sonatsuer) [reply modules/db/test/blaze/db/api_test.clj:477]: Done.

File: modules/db/test/blaze/db/api_test.clj [hunk 5]
```diff
@@ -430,7 +463,45 @@
...
+              :num-changes := 2)))))
+
+    (testing "a patient with several resources referencing it"
+      (satisfies-prop 10
+        (prop/for-all [resources mixed-resource-gen]
+          (with-system-data [{:blaze.db/keys [node]} config]
+            [(into [[:put {:fhir/type :fhir/Patient :id "0"}]
+                    [:put {:fhir/type :fhir/Medication :id "0"}]
+                    [:put {:fhir/type :fhir/Practitioner :id "0"}]]
+                   (map (partial vector :create) resources))]
```
[the remark this answers is not in the document]
Comment (sonatsuer) [reply modules/db/test/blaze/db/api_test.clj]: Done.

File: modules/db/test/blaze/db/api_test.clj [hunk 6]
```diff
@@ -430,7 +463,45 @@
...
+            (let [make-delete-command (fn [{id :id type :fhir/type}] [:delete (name type) id])
+                  delete-list (into [[:delete "Patient" "0"]] (map make-delete-command resources))
```
[the remark this answers is not in the document]
Comment (sonatsuer) [reply modules/db/test/blaze/db/api_test.clj]: Done.

File: modules/db/test/blaze/db/api_test.clj [hunk 7]
```diff
@@ -5455,56 +5577,18 @@
...
-        (prop/for-all [tx-ops tx-ops-gen]
-          (with-system-data [{:blaze.db/keys [node]} config]
-            [tx-ops]
+    (satisfies-prop 10
+      (prop/for-all [resources mixed-resource-gen]
+        (with-system-data [{:blaze.db/keys [node]} config]
+          [(into [[:put {:fhir/type :fhir/Patient :id "0"}]
+                  [:put {:fhir/type :fhir/Medication :id "0"}]
+                  [:put {:fhir/type :fhir/Practitioner :id "0"}]]
+                 (map (partial vector :create) resources))]
```
[the remark this answers is not in the document]
Comment (sonatsuer) [reply modules/db/test/blaze/db/api_test.clj]: Done.

File: modules/db/test/blaze/db/node/tx_indexer/verify_test.clj [hunk 8]
```diff
@@ -381,6 +426,93 @@
...
 
+  (testing "delete conflict"
+    (testing "deleting a patient and creating a resource referencing it in the same transaction"
+      (let [hash (hash/generate observation-0)]
+        (with-system-data [{:blaze.db/keys [node]} config]
+          [[[:put patient-0]]]
+
+          (given (verify/verify-tx-cmds
+                  (d/db node) 2
+                  [{:op "delete" :type "Patient" :id "0" :check-refs true}
```
[the remark this answers is not in the document]
Comment (sonatsuer) [reply modules/db/test/blaze/db/node/tx_indexer/verify_test.clj:433]: Done.

**Pull request merged**